common/browser_process.go

```diff
@@ -179,41 +179,51 @@ func execute(
 	return command{cmd, done, stdout, stderr}, nil
 }
 
-// parseDevToolsURL grabs the websocket address from chrome's output and returns it.
-func parseDevToolsURL(ctx context.Context, cmd command) (wsURL string, _ error) {
-	type result struct {
-		devToolsURL string
-		err         error
-	}
-	c := make(chan result, 1)
+// parseDevToolsURL grabs the WebSocket address from Chrome's output and returns
+// it. If the process ends abruptly, it will return the first error from stderr.
+func parseDevToolsURL(ctx context.Context, cmd command) (string, error) { //nolint: cyclop
+	var (
+		lines = make(chan string)
+		errCh = make(chan error)
+	)
 	go func() {
-		const urlPrefix = "DevTools listening on "
 		scanner := bufio.NewScanner(cmd.stderr)
-
 		for scanner.Scan() {
-			line := scanner.Text()
+			lines <- scanner.Text()
+		}
+		if err := scanner.Err(); err != nil {
+			errCh <- err
+		}
+	}()
+
+	const urlPrefix = "DevTools listening on "
+	var stdErr error
+
+	preferStdErr := func(e error) error {
+		if stdErr != nil {
+			return stdErr
+		}
+		return e
```

Comment:
You could use `defer` instead of calling `preferStdErr` on each case below.

For example:

```go
func parseDevToolsURL(ctx context.Context, cmd command) (devToolsURL string, err error) { //nolint:cyclop
	var (
		lines = make(chan string, 1)
		errCh = make(chan error, 1)
	)
	go func() {
		scanner := bufio.NewScanner(cmd.stderr)
		for scanner.Scan() {
			lines <- scanner.Text()
		}
		if err := scanner.Err(); err != nil {
			errCh <- err
		}
	}()
	var stdErr error
	defer func() {
		if stdErr != nil && devToolsURL == "" {
			err = stdErr
		}
	}()
	for err == nil && devToolsURL == "" {
		select {
		case line := <-lines:
			const urlPrefix = "DevTools listening on "

			if strings.HasPrefix(line, urlPrefix) {
				devToolsURL = strings.TrimPrefix(strings.TrimSpace(line), urlPrefix)
			}
			// TODO: Capture more than one error?
			if stdErr == nil && strings.Contains(line, ":ERROR:") {
				if i := strings.Index(line, "] "); i > 0 {
					stdErr = errors.New(line[i+2:])
				}
			}
		case err = <-errCh:
		case <-cmd.done:
			err = errors.New("browser process ended unexpectedly")
		case <-ctx.Done():
			err = fmt.Errorf("%w", ctx.Err())
		}
	}

	return devToolsURL, err
}
```